ssh/controlmaster.go

```diff
@@ -0,0 +1,233 @@
+		return errors.New("Unable to read entire response packet length")
+	}
+	respLen := binary.BigEndian.Uint32(respLenBuf)
+	// Then read the response
+	resp := make([]byte, respLen)
+	n2, err := c.rw.Read(resp)
+	if err != nil {
+		return err
+	}
+	if uint32(n2) != respLen {
+		return errors.New("Unable to read entire response packet")
+	}
+	if err := Unmarshal(resp, packet); err != nil {
+		return err
+	}
+	return nil
+}
+
+// setupControlMaster gets a CM session setup and switches to proxy mode
+func setupControlMaster(c *controlMasterConnection) error {
+	// Write our own hello with version
+	if err := c.writePacket(&controlMasterHelloMessage{
+		// MUX_MSG_HELLO
+		PacketType:      0x00000001,
+		ProtocolVersion: 4,
+	}); err != nil {
+		return err
+	}
+	// Read the hello message
+	var hello controlMasterHelloMessage
+	if err := c.readPacket(&hello); err != nil {
+		return err
+	}
+	// MUX_MSG_HELLO
+	if hello.PacketType != 0x00000001 {
+		return errors.New("Unknown PacketType, expecting MUX_MSG_HELLO")
+	}
+	// Check that the server sent a protocol we know
+	if hello.ProtocolVersion != 4 {
+		return errors.New("Unknown ControlMaster protocol")
+	}
+	// Send an alive check with request id 1
+	if err := c.writePacket(&controlMasterMessage{
+		// MUX_C_ALIVE_CHECK
+		PacketType: 0x10000004,
+		RequestID:  1,
+	}); err != nil {
+		return err
+	}
+	// Read the alive check response message
+	var alive controlMasterAliveMessage
+	if err := c.readPacket(&alive); err != nil {
+		return err
+	}
+	// MUX_S_ALIVE
+	if alive.PacketType != 0x80000005 {
+		return errors.New("Unknown PacketType, expecting MUX_C_ALIVE_CHECK")
+	}
+	// Request that we move to proxy mode with request id 2
+	if err := c.writePacket(&controlMasterMessage{
+		// MUX_C_PROXY
+		PacketType: 0x1000000F,
+		RequestID:  2,
+	}); err != nil {
+		return err
+	}
+	// Wait for confirmation we are now in proxy mode
+	var proxy controlMasterMessage
+	if err := c.readPacket(&proxy); err != nil {
+		return err
+	}
+	// MUX_S_PROXY
+	if proxy.PacketType != 0x8000000F {
+		return errors.New("Unknown PacketType, expecting MUX_S_PROXY")
+	}
+	return nil
+}
+
+// controlMasterProxyConnection implements the controlmaster proxy protocol
+type controlMasterProxyConnection struct {
+	rwc io.ReadWriteCloser
+}
+
+// writePacket prefixes the length and a 0 byte of padding
+func (c *controlMasterProxyConnection) writePacket(packet []byte) error {
+	// Packet length is the length of data plus 1 byte of padding
+	packetLen := uint32(len(packet) + 1)
+	msgLenBytes := make([]byte, 5)
+	binary.BigEndian.PutUint32(msgLenBytes[0:4], packetLen)
+	packet = append(msgLenBytes, packet...)
+	n, err := c.rwc.Write(packet)
+	if err != nil {
+		return err
+	}
+	if n != len(packet) {
+		return errors.New("Unable to write all bytes in proxy packet")
+	}
+	return nil
+}
+
+// readPacket reads a packet based on an initial length value minus padding
+func (c *controlMasterProxyConnection) readPacket() ([]byte, error) {
```

Comment:
Do these need to be safe for concurrent use?  If so, you need a mutex to serialize reading... I can't quite tell whether this is expected or not.